hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/mongodb/FieldResolver.java

```diff
@@ -0,0 +1,227 @@
+import static com.hazelcast.jet.sql.impl.connector.mongodb.BsonTypes.resolveTypeFromJava;
+import static com.hazelcast.jet.sql.impl.connector.mongodb.Options.DATA_LINK_REF_OPTION;
+import static com.hazelcast.sql.impl.type.QueryDataType.VARCHAR;
+import static com.mongodb.client.model.Filters.eq;
+import static java.util.Objects.requireNonNull;
+
+class FieldResolver {
+
+    /**
+     * Resolves fields based on the options passed to the connector and schema provided by user (if any)
+     * @param options options provided for the connector
+     * @param userFields user-defined field list
+     * @return resolved fields - all fields from collection or user provided fields with resolved types
+     *
+     * @throws IllegalArgumentException if given column type won't be resolved or field from user mapping won't exist
+     *                                  in the collection
+     */
+    List<MappingField> resolveFields(
+            @Nonnull Map<String, String> options,
+            @Nonnull List<MappingField> userFields,
+            boolean stream
+    ) {
+        Map<String, DocumentField> dbFields = readFields(options, stream);
+
+        List<MappingField> resolvedFields = new ArrayList<>();
+        if (userFields.isEmpty()) {
+            for (DocumentField documentField : dbFields.values()) {
+                try {
+                    MappingField mappingField = new MappingField(
+                            documentField.columnName,
+                            resolveType(documentField.columnType),
+                            documentField.columnName
+                    );
+                    mappingField.setPrimaryKey(isId(documentField.columnName, stream));
+                    resolvedFields.add(mappingField);
+                } catch (IllegalArgumentException e) {
+                    throw new IllegalArgumentException("Could not load column class " + documentField.columnType, e);
+                }
+            }
+        } else {
+            for (MappingField f : userFields) {
+                String prefixIfStream = stream ? "fullDocument." : "";
+                String nameInMongo = f.externalName() == null ? prefixIfStream + f.name() : f.externalName();
+
+                DocumentField documentField = dbFields.get(nameInMongo);
+                if (documentField == null) {
+                    throw new IllegalArgumentException("Could not resolve field with name " + nameInMongo);
+                }
+                MappingField mappingField = new MappingField(f.name(), f.type(), nameInMongo);
+                mappingField.setPrimaryKey(isId(documentField.columnName, stream));
+                validateType(f, documentField);
+                resolvedFields.add(mappingField);
+            }
+        }
+        return resolvedFields;
+    }
+
+    boolean isId(String nameInMongo, boolean stream) {
+        if (stream) {
+            return "fullDocument._id".equalsIgnoreCase(nameInMongo);
```

Comment:
We need to have `_id` mapping in streaming. I couldn't find any particular reason, just otherwise queries were not executing correctly and adding explicit `_id` field to the projection was suggested